03-dplyr.Rmd

```diff
@@ -548,140 +541,133 @@ surveys_gw <- surveys %>%
-and then gathering can be a useful way to balance out a dataset so every
-replicate has the same composition.
+Note that now the `NA` genera are included in the re-gathered format. Pivoting wider and then longer can be a useful way to balance out a dataset so every replicate has the same composition.
 
-We could also have used a specification for what columns to include. This can be
-useful if you have a large number of identifying columns, and it's easier to
-specify what to gather than what to leave alone. And if the columns are directly
-adjacent, we don't even need to list them all out - just use the `:` operator!
+We could also have used a specification for what columns to omit from reshaping. In this example, we will use all columns _except_ `plot_id` for the names variable. By using the minus sign in the `cols` argument, we exclude `plot_id` from being reshaped.
 
 ```{r, purl=FALSE}
-surveys_spread %>%
-  gather(key = "genus", value = "mean_weight", Baiomys:Spermophilus) %>%
+surveys_wide %>%
+  pivot_longer(names_to = "genus", values_to = "mean_weight", cols = -plot_id) %>%
   head()
 ```
 
 > ### Challenge {.challenge}
 >
-> 1. Spread the `surveys` data frame with `year` as columns, `plot_id` 
+> 1. Reshape the `surveys` data frame with `year` as columns, `plot_id` 
 >   as rows, and the
 >   number of genera per plot as the values. You will need to summarize before
 >   reshaping, and use the function `n_distinct()` to get the number of unique
 >   genera within a particular chunk of data. It's a powerful function! See
 >   `?n_distinct` for more.
 > 
-> ```{r, answer=TRUE, purl=FALSE}
-> surveys_spread_genera <- surveys %>%
+>```{r, answer=TRUE, purl=FALSE}
+>surveys_wide_genera <- surveys %>%
 >   group_by(plot_id, year) %>%
 >   summarize(n_genera = n_distinct(genus)) %>%
->   spread(year, n_genera)
+>   pivot_wider(names_from = year, values_from = n_genera)
 > 
-> head(surveys_spread_genera)
-> ```
+>head(surveys_wide_genera)
+>```
 >
-> 2. Now take that data frame and `gather()` it again, so each row is a unique
+> 2. Now take that data frame and `pivot_longer()` it again, so each row is a unique
 >    `plot_id` by `year` combination.
 >
 > ```{r, answer=TRUE, purl=FALSE}
-> surveys_spread_genera %>%
->   gather("year", "n_genera", -plot_id)
+> surveys_wide_genera %>%
+>   pivot_longer(names_to = "year", values_to = "n_genera", cols = -plot_id)
 > ```
 >
 > 3. The `surveys` data set has
 >    two measurement columns: `hindfoot_length` and `weight`.  This makes it
 >    difficult to do things like look at the relationship between mean values of
 >    each measurement per year in different plot types. Let's walk through a
->    common solution for this type of problem. First, use `gather()` to create a
+>    common solution for this type of problem. First, use `pivot_longer()` to create a
 >     dataset where we have a key column called `measurement` and a
```

Comment:
Maybe it would be clearer to write something like

> There is a column called `measurement` that contains the type of measurement (hindfoot length or `weight) ...